src/emailcontactlistmodel.cpp

```diff
@@ -0,0 +1,64 @@
+#include "emailcontactlistmodel.h"
+#include "emailcontactstorage.h"
+
+
+EmailContactListModel::EmailContactListModel(QAbstractItemModel *parent)
+    : QAbstractListModel(parent)
+{    
+}
+
+
+QHash<int, QByteArray> EmailContactListModel::roleNames() const
+{
+    static QHash<int, QByteArray> names = {
+        {NameRole, "name"},
+        {EmailRole, "email"},
+        {TimeStampRole, "timestamp"}
+    };
+    return names;
+}
+
+
+int EmailContactListModel::rowCount(const QModelIndex &parent) const
+{
+    Q_UNUSED(parent)
+    return EmailContactStorage::instance()->count();
+}
+
+
+QVariant EmailContactListModel::data(const QModelIndex &index, int role) const
+{
+    EmailContactStorage *storage = EmailContactStorage::instance();
+    if (!index.isValid() || storage == nullptr || index.row() > rowCount())
+        return QVariant();
+
+    const EmailContact &contact = storage->at(index.row());
+
+    switch (role) {
+    case NameRole:
+        return contact.name();
+    case EmailRole:
+        return contact.email();
+    case TimeStampRole:
+        return contact.timeStamp().toLocalTime();
+    }
+    return QVariant();
+}
+
+
+void EmailContactListModel::removeDuplicates(QAbstractItemModel *model, const int role)
+{
+    if (model == nullptr || model->rowCount() < 0 || !model->roleNames().contains(role))
+        return;
+
+    EmailContactStorage *storage = EmailContactStorage::instance();
+
+    for (int i = 0; i != model->rowCount(); ++i) {
+        QModelIndex index = model->index(i, 0);
+        QVariant variant = model->data(index, role);
+
+        QString name = variant.toString();
+        if (!name.isEmpty())
+            storage->removeByName(name);
```

Comment:
This doesn't either play well with a common storage of the contacts. If it's allowed to remove duplicates (or why should there be duplicates ever?) those should be removed only from a single instance?

Reply:
Possibly a bad name.
Duplicates from another model that we don't know anything about in this class.